pkg/chunkenc/memchunk.go

```diff
@@ -229,7 +231,8 @@ func NewByteChunk(b []byte) (*MemChunk, error) {
 		// Verify checksums.
 		expCRC := binary.BigEndian.Uint32(b[blk.offset+l:])
 		if expCRC != crc32.Checksum(blk.b, castagnoliTable) {
-			return bc, ErrInvalidChecksum
+			level.Error(util.Logger).Log("msg", "Checksum does not match for a block in chunk, this block will be skipped", "err", ErrInvalidChecksum)
```

Comment:
We don't have any indication to blocks in a chunk. We can identify the blocks only with their offsets. Can we add a field `block number` to `block` struct so that we can indicate blocks uniquely?

Reply:
I wouldn't change anything here. I think this looks good.

However if the full chunk is invalid it still send an error see above :
```go
if expCRC != db.crc32() {
		return nil, ErrInvalidChecksum
	}
```

which means we need to take care of this.

The error will buble up from cortex to this line https://github.com/grafana/loki/blob/master/pkg/storage/iterator.go#L404

Careful this is difficult to change :)

But you want to:

1 - Check if `err == chunkenc.ErrInvalidChecksum` if so you need to skip that chunk and log the error. (meaning send back `errChan <- nil`)
2 - However you can't leave an item in `chunks []*chunkenc.LazyChunk` with `LazyChunk.Chunk.Data` empty (otherwise you'll endup with an iterator error here https://github.com/grafana/loki/blob/master/pkg/chunkenc/lazy_chunk.go#L28), this means you need to remove the item from the list. I think you should do that after waiting for all err from the chan :
```go
	var lastErr error
	for i := 0; i < len(chksByFetcher); i++ {
		if err := <-errChan; err != nil {
			lastErr = err
		}
	}
if lastErr == nil {
// remove any lazy chunk with no data aka chunks[i].Chunk.Data == nil.
}
```

I think it's key here that you write a good test for this function to ensure it skips correctly bad chunks. You can abstract away the Fetcher interface and fake this error. `chks, err := fetcher.FetchChunks(ctx, chks, keys)` since fetcher is passed inside LazyChunk array.

Take your time to understand this function and good luck ✌️ .

Reply:
No need for now. Let's first tackle the problem of not being able to reply if couple of chunks are bad out of the whole batch.

Reply:
@cyriltovena should we just check for `invalid checksum`. Any error which comes here https://github.com/cortexproject/cortex/blob/f1f513ec1de84e87ede492cad87ed8e6d2af0237/pkg/chunk/local/fs_object_client.go#L77 is a potential chunk error. The `invalid checksum` error may we come here https://github.com/cortexproject/cortex/blob/f1f513ec1de84e87ede492cad87ed8e6d2af0237/pkg/chunk/local/fs_object_client.go#L92 due to `invalid checksum` of chunk meta. Later in the code `NewByteChunk` of `loki` code is called and we might get `invalid checksum` there  as well.

So, my question is, should we just check `invalid checksum` error given by `loki` code and just log it and continue iterating other chunks?
OR
Should we  just log any error related `chunk` rather than returning that error?

Reply:
I would continue only in case of `invalid checksum`, this won't resolved itself. Other errors can be logged but we should stop.

Reply:
Ok. Will update code accordingly